File: Python/81.Search in Rotated Sorted Array II.py
#在有序循环数组中寻找目标数（有重复的数)
class Solution(object):
    def search(self, nums, target):
        l , r = 0 , len(nums)-1
        while l<=r:
            if nums[l]==target:
                return True
            if nums[r]==target:
                return True
            if nums[l]<target and nums[r]>target:
                while l<=r:
                    mid = (r+l)//2
                    if nums[mid]==target:
                        return True
                    elif nums[mid]<target:
                        l = mid+1
                    else:
                        r = mid-1
                return False
            else:
                l += 1
                r -= 1
        return False

s = Solution()
print(s.search([1,1,1,2,2,3,3,3],2))

File: C++/Array/81.Search in Rotated Sorted Array II.cpp
/*
有序数组（可能有重复数字），左右移动某些位，在数组中寻找给定数。
*/
/*
思路：先搞清楚数组当前结构，然后用给定数判断其位置。
重复数字的影响？连续相等数字，造成左右混淆。
[1,1,1,3,1],[1,3,1,1,1]不清楚3在中间数的左侧还是右侧
 */
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
	bool search(vector<int>& nums, int target) {
		int length = nums.size();
		if (length == 0)	return false;
		int left = 0, right = length - 1, mid;
		while(left < right)
		{
			mid = (left+right)/2;
			if(target==nums[mid])	return true;
			if(nums[left]>nums[mid])
			{
				if(target>=nums[left] || target<nums[mid])	right = mid-1;
				else left = mid+1;
			}
			else if(nums[left]<nums[mid])
			{
				if(target>nums[mid] || target<nums[left])	left = mid+1;
				else right = mid-1;
			}
			else
			{
				/*先没必要一个一个找，先挪一个，如果可以比较出来大小，就可以继续二分。
				for(int i = left; i<=right;++i)
				{
					if(nums[i]==target)	return true;
				}
				return false;*/
				++left;
			}
		}
		return (nums[left]==target)?true:false;
	}
};

int main()
{
	Solution s;
	//vector<int> nums{ 4, 5, 6, 7, 0, 0 , 1, 2, 3 };
	vector<int> nums{ 1, 3 , 1 , 1 , 1};
	cout<<s.search(nums, 3)<<endl;
}

